File: example/1d-linear-convection/weno3/python/05/reconstructor/factory.py
# reconstructor/factory.py
from .eno import EnoReconstructor
from .weno3 import Weno3Reconstructor

class ReconstructorFactory:
    _schemes = {
        "eno": EnoReconstructor,
        "weno3": Weno3Reconstructor,      # ← 注意：这里用 "weno3"
        # "weno5": Weno5Reconstructor,   # 未来扩展
    }

    @staticmethod
    def create(config, domain):
        scheme = config.recon_scheme.lower()
        order = getattr(config, 'spatial_order', None)
        
        # ✅ 关键：将 "weno" 自动映射为 "weno3", "weno5" 等
        if scheme == "weno":
            if order is None:
                raise ValueError("使用 'weno' 时必须设置 config.spatial_order")
            scheme = f"weno{order}"

        # 检查是否支持
        if scheme not in ReconstructorFactory._schemes:
            supported = list(ReconstructorFactory._schemes.keys())
            raise ValueError(f"不支持的重建格式：'{scheme}'（支持：{supported}）")

        recon_cls = ReconstructorFactory._schemes[scheme]

        # 根据 scheme 类型创建实例
        if scheme.startswith("eno"):
            return recon_cls(order, domain.ntcells)
        elif scheme.startswith("weno"):
            return recon_cls()  # WENO 类通常无参
        # 可扩展 elif...

File: example/1d-linear-convection/weno3/julia/03f/src/reconstructor/factory.jl
# src/reconstructor/factory.jl

module ReconstructorFactory

const _REGISTRY = Dict{String, Function}()

function register(name::String, ctor::Function)
    if haskey(_REGISTRY, name)
        @warn "重构器 '$name' 已注册，将被覆盖"
    end
    _REGISTRY[name] = ctor
end

function create(config::Any, domain::Any)
    scheme = lowercase(string(getfield_safe(config, :recon_scheme, "")))
    if scheme == ""
        error("必须先通过 with_reconstruction 设置重建格式！")
    end

    # 处理 WENO 默认命名
    if scheme == "weno"
        order = getfield_safe(config, :spatial_order, 5)
        scheme = "weno$(order)"
    end

    if !haskey(_REGISTRY, scheme)
        available = sort(collect(keys(_REGISTRY)))
        error("未注册的重构器: '$scheme'。可用选项: $available")
    end

    return _REGISTRY[scheme](config, domain)
end

# ✅ 手动注册（与你原有风格一致）
register("eno", (config, domain) -> Main.EnoReconstructor(config.spatial_order, domain.ntcells))
register("weno3", (config, domain) -> Main.Weno3Reconstructor())
register("weno5", (config, domain) -> Main.Weno5Reconstructor())  # ← 新增

# ---------------------- 辅助函数 ----------------------
function getfield_safe(obj, name::Symbol, default)
    return isdefined(obj, name) ? getfield(obj, name) : default
end

end  # module ReconstructorFactory

export ReconstructorFactory